hazelcast/src/test/java/com/hazelcast/client/partitionservice/ClientMigrationListenerTest.java

```diff
@@ -0,0 +1,269 @@
+
+        HazelcastInstance instance1 = hazelcastFactory.newHazelcastInstance();
+        HazelcastInstance client = clientFactory.apply(clientListener);
+        warmUpPartitions(instance1, client);
+
+        // Change to NO_MIGRATION to prevent repartitioning
+        // before 2nd member started and ready.
+        instance1.getCluster().changeClusterState(ClusterState.NO_MIGRATION);
+
+        HazelcastInstance instance2 = hazelcastFactory.newHazelcastInstance();
+
+        changeClusterStateEventually(instance2, ClusterState.ACTIVE);
+        waitAllForSafeState(instance2, instance1, client);
+
+        assertRegistrationsSizeEventually(instance1, 1);
+        assertFunction.accept(clientListener);
+    }
+
+    private void verifyMigrationListenerNeverInvoked(MigrationListener listener) {
+        verify(listener, never()).migrationStarted(any(MigrationStateImpl.class));
+        verify(listener, never()).migrationFinished(any(MigrationStateImpl.class));
+        verify(listener, never()).replicaMigrationCompleted(any(ReplicaMigrationEvent.class));
+        verify(listener, never()).replicaMigrationFailed(any(ReplicaMigrationEvent.class));
+    }
+
+    private void assertMigrationProcess(PartitionMigrationListenerTest.EventCollectingMigrationListener listener) {
+        PartitionMigrationListenerTest.MigrationEventsPack eventsPack = listener.ensureAndGetSingleEventPack();
+        assertMigrationProcessCompleted(eventsPack);
+        assertMigrationProcessEventsConsistent(eventsPack);
+        assertMigrationEventsConsistentWithResult(eventsPack);
+    }
+
+    private PartitionMigrationListenerTest.EventCollectingMigrationListener eventCollectingMigrationListener() {
+        return new PartitionMigrationListenerTest.EventCollectingMigrationListener();
+    }
+
+    private void assertRegistrationsSizeEventually(HazelcastInstance instance, int size) {
+        assertTrueEventually(() -> {
+            EventService eventService = getNode(instance).getNodeEngine().getEventService();
+            Collection<EventRegistration> registrations = eventService
+                    .getRegistrations(SERVICE_NAME, MIGRATION_EVENT_TOPIC);
+            assertEquals(size, registrations.size());
+        });
+    }
+
+    static class SingleThreadMigrationListener implements MigrationListener {
+
+        private String threadName;
+        private boolean finished = false;
+
+        @Override
+        public void migrationStarted(MigrationState state) {
+            assertNull(threadName);
+            threadName = Thread.currentThread().getName();
+        }
+
+        @Override
+        public void migrationFinished(MigrationState state) {
+            assertNotNull(threadName);
+            assertEquals(threadName, Thread.currentThread().getName());
+            finished = true;
+        }
+
+        @Override
+        public void replicaMigrationCompleted(ReplicaMigrationEvent event) {
+            assertNotNull(threadName);
```

Comment:
This and other asserts in this and `replicaMigrationFailed` methods don't affect the test result, i.e. if we change this line to `assertNull(threadName);`, the test still passes. Maybe, it's a good idea to introduce another boolean field which will be changed to `false` when the listener is invoked on a different thread. Then this field could be checked in `assertAllMethodsInvokedOnTheSameThread` method.